stdlib/@tests/test_cases/builtins/check_slice.py

```diff
@@ -0,0 +1,159 @@
+
+t8: "slice[int, int]" = slice(1, None)
+t9: "slice[int, int]" = slice(1, None, None)
+t10: "slice[int, int]" = slice(1, None, "foo")
+
+t11: "slice[int, int]" = slice(1, 1)
+t12: "slice[int, int]" = slice(1, 1, None)
+t13: "slice[int, int]" = slice(1, 1, "foo")
+# endregion Tests for slice[X, Y] assignments ------------------------------------------
+
+# region Tests for slice[X, Y, Z] assignments ------------------------------------------
+u0: "slice[int, int, int]" = slice(None)
+u1: "slice[int, int, int]" = slice(None, None)
+u2: "slice[int, int, int]" = slice(None, None, None)
+u3: "slice[int, int, int]" = slice(None, None, 1)
+
+u4: "slice[int, int, int]" = slice(1)
+u5: "slice[int, int, int]" = slice(None, 1)
+u6: "slice[int, int, int]" = slice(None, 1, None)
+u7: "slice[int, int, int]" = slice(None, 1, 1)
+
+u8: "slice[int, int, int]" = slice(1, None)
+u9: "slice[int, int, int]" = slice(1, None, None)
+u10: "slice[int, int, int]" = slice(1, None, 1)
+
+u11: "slice[int, int, int]" = slice(1, 1)
+u12: "slice[int, int, int]" = slice(1, 1, None)
+u13: "slice[int, int, int]" = slice(1, 1, 1)
+# endregion Tests for slice[X, Y, Z] assignments ---------------------------------------
+
+# region Tests for slice properties ----------------------------------------------------
+assert_type(slice(1).stop, int)
+assert_type(slice(None, 1).stop, int)
+assert_type(slice(None, 1, None).stop, int)
+
+assert_type(slice(1, None).start, int)
+assert_type(slice(1, None, None).start, int)
+
+assert_type(slice(None, None, 1).step, int)
+# endregion Tests for slice properties -------------------------------------------------
+
+
+# region Integration tests for slices with datetimes -----------------------------------
+class TimeSeries:  # similar to pandas.Series with datetime index
+    def __getitem__(self, key: "slice[DT | str | None, DT | str | None]") -> Any:
+        """Subsample the time series at the given dates."""
+        ...
+
+
+class TimeSeriesInterpolator:  # similar to pandas.Series with datetime index
+    def __getitem__(self, key: "slice[DT, DT, TD | None]") -> Any:
+        """Subsample the time series at the given dates."""
+        ...
+
+
+# tests slices as an argument
+start = DT(2021, 1, 1)
+stop = DT(2021, 1, 10)
+step = TD(days=1)
+# see: https://pandas.pydata.org/docs/user_guide/timeseries.html#partial-string-indexing
+series = TimeSeries()
+_ = series[None:"2022-01-10"]
+_ = series["2022-01-01":None]
+_ = series["2022-01-01":"2022-01-10"]
+_ = series[None:stop]
+_ = series[start:None]
+_ = series[start:stop]
+_ = series[:]
+model = TimeSeriesInterpolator()
+_ = model[start:stop]
+_ = model[start:stop:step]
+_ = model[start:stop:None]
+
+
+# test slices as a return type
+def foo(flag: bool, value: DT) -> "slice[DT, None] | slice[None, DT]":
+    if flag:
+        return slice(value, None)  # slice[DT, DT|Any, Any] incompatible
+    else:
+        return slice(None, value)  # slice[DT|Any, DT, Any] incompatible
```

Comment:
These tests doesn't seem to pass with mypy or pyright currently

Reply:
Oh, sorry, I see you already noted this in https://github.com/python/typeshed/pull/12904#discussion_r1838201403. I'll continue the conversation there 👍